### PR TITLE
EECS 149 applications ported by @gundralaa

### PR DESCRIPTION
**Pull request by lhstrh**
Based on https://github.com/gundralaa/lf-buckler which has `CRLF` line endings, which messes up the diff in #1.

### CODE REVIEW

**Comment by lsk567** [comment]
Thanks for taking a look at the diff issue! @lhstrh

File: scripts/build_nrf_unix.sh [hunk 1]
```diff
@@ -3,12 +3,19 @@
...
 
+
 echo "starting NRF generation script into $LF_SOURCE_GEN_DIRECTORY"
 echo "pwd is $(pwd)"
 
 cp $PROJECT_ROOT/platform/lf_nRF52832_support.c $LF_SOURCE_GEN_DIRECTORY/core/platform/
 cp $PROJECT_ROOT/platform/lf_nRF52832_support.h $LF_SOURCE_GEN_DIRECTORY/core/platform/
 cp $PROJECT_ROOT/platform/platform.h $LF_SOURCE_GEN_DIRECTORY/core/
+
+cp $PROJECT_ROOT/platform/lf_nRF52832_support.c $LF_SOURCE_GEN_DIRECTORY/include/core/platform/
```
Comment (edwardalee) [inline scripts/build_nrf_unix.sh]: Would it be better to create a symbolic link from include/core to core?

File: src/KobukiBLE.lf [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    }
+}
+
+main reactor {
+    timer t(0, 10 msec);
+    robo = new Robot();
+    reaction(t) -> robo.clck {=
+        SET(robo.clck, 1);
+    =}
+}
```
Comment (edwardalee) [inline src/KobukiBLE.lf]: Add trailing newline?

File: src/KobukiSensor.lf [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+}
+
+main reactor {
+    timer t(1 sec, 500 msec);
+    robot = new Robot();
+    reaction(t) -> robot.clck {=
+        SET(robot.clck, 1);
+    =}
+}
```
Comment (edwardalee) [inline src/KobukiSensor.lf]: Add trailing sensor?

**Review by edwardalee — CHANGES_REQUESTED** [review]
This looks great! However, there is a lot of code duplication among the .lf files.  Could you create one or more base classes, e.g. for the Robot, so that the code doesn't have to be repeated in each application? See https://www.lf-lang.org/docs/handbook/extending-reactors

File: platform/lf_nrf52_support.h [hunk 4]
```diff
@@ -0,0 +1,215 @@
+<<<<<<< Updated upstream:platform/lf_nRF52832_support.h
```
Comment (lhstrh) [inline platform/lf_nrf52_support.h]: Whoops?

File: platform/lf_nrf52_support.h [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+ * Text descriptions and spaces add an additional 55,
+ * for a total of 79. One more allows for a null terminator.
+ */
+#define LF_TIME_BUFFER_LENGTH 80
+
+
+// The underlying physical clock for Linux
+#define _LF_CLOCK CLOCK_MONOTONIC
+
+=======
```
Comment (lhstrh) [inline platform/lf_nrf52_support.h]: Conflict was not resolved (and I bet CRLF line endings were reintroduced)

File: scripts/ble_scripts/ble_utils.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        raise ValueError("Invalid address supplied")
+
+    return args
+
+def handle_sigint():
+    def signal_handler(signal, frame):
+        sys.exit(0)
+    signal.signal(signal.SIGINT, signal_handler)
+
+LAB11 = 0x02e0
```
Comment (lhstrh) [inline scripts/ble_scripts/ble_utils.py]:
```suggestion
LAB11 = 0x02e0

```

File: scripts/ble_scripts/discover.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import asyncio
+from bleak import BleakScanner
+
+async def main():
+    devices = await BleakScanner.discover()
+    for d in devices:
+        print(d)
+
+asyncio.run(main())
```
Comment (lhstrh) [inline scripts/ble_scripts/discover.py]:
```suggestion
asyncio.run(main())

```

File: scripts/ble_scripts/led_writer.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+from bleak import BleakClient, BleakError
+
+from ble_utils import parse_ble_args, handle_sigint, LAB11
+args = parse_ble_args('Communicates with buckler LED toggle characteristic')
+addr = args.addr.lower()
+timeout = args.timeout
+handle_sigint()
+
+# TODO: Make sure these UUIDs match the ones in the LF files.
```
Comment (lhstrh) [inline scripts/ble_scripts/led_writer.py]: @gundralaa: Could you elaborate on this `TODO`?

File: src/KobukiBLE.lf [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    }
+}
+
+main reactor {
+    timer t(1 sec, 10 msec);
+    robo = new Robot();
+    reaction(t) -> robo.clck {=
+        SET(robo.clck, 1);
+    =}
+}
```
Comment (lhstrh) [inline src/KobukiBLE.lf]:
```suggestion
}

```

File: src/KobukiSensor.lf (hunk 3 above)
Comment (lhstrh) [inline src/KobukiSensor.lf]:
```suggestion
}

```

File: src/KobukiSquare.lf [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+
+
+main reactor {
+    timer t(1 sec, 10 msec);
+    robot = new Robot();
+    reaction(t) -> robot.clck {=
+        SET(robot.clck, 1);
+    =}
+}
```
Comment (lhstrh) [inline src/KobukiSquare.lf]:
```suggestion
}

```

File: src/KobukiSquareIMU.lf [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+main reactor {
+    // TODO: here we wait arbitrary till init complete
+    // ensures first poll occurs after init
+    timer t(1 sec, 10 msec);
+    robot = new Robot();
+    reaction(t) -> robot.clck {=
+        SET(robot.clck, 1);
+    =}
+}
```
Comment (lhstrh) [inline src/KobukiSquareIMU.lf]:
```suggestion
}

```

File: src/KobukiTemplate.lf [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+  
+}
+
+main reactor {
+    timer t(1 sec, 10 msec);
+    robot = new Robot();
+    reaction(t) -> robot.clck {=
+        SET(robot.clck, 1);
+    =}
+}
```
Comment (lhstrh) [inline src/KobukiTemplate.lf]:
```suggestion
}

```

File: src/ModalRomi.lf [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    initialize = new Init();
+    robot = new Robot();
+    reaction(initialize.err_out) {=
+        self->err = initialize.err_out->value;
+    =}
+    reaction(t) -> robot.clck {=
+        SET(robot.clck, self->err);
+    =}
+
+}
```
Comment (lhstrh) [inline src/ModalRomi.lf]:
```suggestion
}

```

File: src/RobotInit.lf [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        APP_ERROR_CHECK(error_code);
+        lsm9ds1_init(&twi_mngr_instance);
+        printf("IMU initialized!\n");
+
+        // initialize Kobuki
+        kobukiInit();
+        printf("Kobuki initialized!\n");
+        SET(err_out, error_code);
+    =}
+}
```
Comment (lhstrh) [inline src/RobotInit.lf]:
```suggestion
}

```

File: src/RomiTemplate/controller.lf [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    r_sensor = new Sensor();
+    r_states = new States();
+
+    clck -> r_sensor.clck;
+    clck -> r_states.clck;
+    
+    r_sensor.btn -> r_states.btn;
+    r_sensor.dist -> r_states.dist;
+    r_states.reset -> r_sensor.reset;
+}
```
Comment (lhstrh) [inline src/RomiTemplate/controller.lf]:
```suggestion
}

```

File: src/RomiTemplate/main.lf [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+}
+
+main reactor {
+    loop = new Loop();
+    control = new Controller();
+
+    loop.clck -> control.clck;
+
+}
```
Comment (lhstrh) [inline src/RomiTemplate/main.lf]:
```suggestion
}

```

File: src/SerialTest.lf [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+        kobukiSensorPoll(&sensors);
+        self->toggle ^= is_button_pressed(&sensors);
+        if (self->toggle) {
+            display_write("ON", DISPLAY_LINE_0);
+        } else {
+            display_write("OFF", DISPLAY_LINE_0);
+        }
+    =}
+
+}
```
Comment (lhstrh) [inline src/SerialTest.lf]:
```suggestion
}

```

File: src/SimpleBLE_adv.lf [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    }
+}
+
+main reactor {
+    timer t(0, 50 msec);
+    board = new Board();
+    reaction(t) -> board.clck {=
+        SET(board.clck, 1);
+    =}
+}
```
Comment (lhstrh) [inline src/SimpleBLE_adv.lf]:
```suggestion
}

```

File: src/SimpleBLE_serv.lf [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+        
+    =}
+    **/
+    
+    timer t(1 sec, 10 msec);
+    board = new Board();
+    reaction(t) -> board.clck {=
+        SET(board.clck, 1);
+    =}
+}
```
Comment (lhstrh) [inline src/SimpleBLE_serv.lf]:
```suggestion
}

```

**Review by lhstrh — COMMENTED** [review]
This is looking great, @gundralaa. I plan to toy with this tomorrow morning. Would you be available for help if I get stuck?

**Pull request merged**